### PR TITLE
Make OpenNIR installable

### PR DESCRIPTION
**Pull request by cash**
I'd like to be to pip install OpenNIR and use components of it. This is a work in progress to get feedback.

The main issues right now are the files in the bin directory (2 jars, a compiled C executable, and a perl script).

I added a dependency on pyserini to get access to the anserini jar that is packaged with it. I looked into using pyserini directly, but it does not have the log watcher that you built for indexing. Its only API for building an index is using string command line arguments that get passed to a main method. (Refs #26)

I did not include the lucene backward compatibility codec jar since those classes are included in the anserini fat jar. 

If this looks like a good approach to you, I'll move on to the perl script and the trec_eval executable.

### CODE REVIEW

**Comment by seanmacavaney** [comment]
Thanks for this work Cash!

One thing I need to do is look into whether there are any customizations made to the current Anserini JAR-- I know for sure prior versions did (e.g., to support other language analyzers for indexing and retrieval), but I may have removed most of them (hopefully all of them!) as Anserini added more of the features I needed.

The dependencies pyserini has now are a bit annoying (tensorflow, faiss, transformers, etc.), but ONIR itself isn't exactly lightweight as far as dependencies go. I mostly worry about transformers and possible headaches that will bring with the million incompatible versions they have there. Perhaps be a good catalyst to bring all that up to their 4.x releases (which HGF promises is going to be stable from now on). The one good thing about the situation is that they changed the package name several times, so we can continue to use some of the older versions without incompatibilities.

[jtreceval](https://github.com/terrierteam/jtreceval) may be a good option for replacing trec_eval. It's a fully-featured cross-platform wrapper of trec_eval, and since we already have pyjnius, it shouldn't be too hard to leverage. Unfortunately, [pytrec_eval](https://github.com/cvangysel/pytrec_eval) doesn't yet support everything we need from trec_eval, so that's not a full replacement at this time. gdeval should only be used for err@x calculations. I think we can live with just a fatal error message on platforms that somehow do not have perl.

The configuration files may also be a challenge? Will we need a way to specify when the configuration should come from those included in the package vs in a local directory? Maybe without any prefix comes from the package (e.g., config/epic), and with a prefix comes from working directory (./config/epic or ~/something/config/epic).

**Comment by cash** [comment]
I agree on feeling uncomfortable with the number of dependencies that pyserini requires. The other option is to include your customized jar in the package just like anserini does. Could do the same with the perl script and as you said detect when it fails and not provide those stats. trec_eval is a little harder since it is compiled code.

I was tempted to punt on the configs and not include them in the installed package. The user would need the whole config tree in their local directory. I suppose they could be included in the package as a starting point and there would be a command that initializes a directory by copying them over. The user can then edit/add as needed for their experiments. The other option is working out some way to handle combining local configs and the package configs as you point out. I'm not familiar enough with the code yet to comment on that.

**Comment by cash** [comment]
I've updated the PR to use the jar in the repository. Needed to move the files into the onir module. The code now figures out the path to those files when running. I'm looking at the config next to see what happens when running as an installed package rather than in the local directory.

**Comment by seanmacavaney** [comment]
Great, thanks for the contributions!

**Comment by cash** [comment]
I can now pip install the OpenNIR package and run the onir_pipeline script that gets installed. I then create a clean directory and copy the config directory into it. Activate the virtual environment where I installed OpenNIR and then run the script from that directory. I cannot run the scripts anywhere that doesn't have a config directory because of this line:

```
result.update(_parse_args(['./config']))
```
There are no config files being installed into the onir module.

Options:
1. Leave it like this. Maybe add a command that generates the default config files in the current directory to give people a starting point.
2. Extend the current code so that there is an internal default config in the module and a user can also pass the path to additional config. The config code would use the internal config as a fallback. There are variations on this where a user can select between internal or external config.

My assumption is that there are two classes of users: those who want to reproduce a result and those who want to run their own experiments. Those who want to reproduce a result would not be changing any configuration so an internal config works for them. The other users will be editing the config and it is probably easier to have the config in a single place rather than two sets of files.

I think prefer option 1. Thoughts?

For reference, I tested with this branch and used the command `onir_pipeline config/conv_knrm config/antique` to verify that it worked.

**Comment by seanmacavaney** [comment]
I think it's worth considering what the use cases are for installing as a package vs cloning the repository. It seems reasonable to me to have these two modes:

As a package:
 - Using components supplied by onir in other code.

From source:
 - Development/experimentation with the existing pipeline (e.g., adding new rankers, datasets, vocabs, etc.)
 - Reproducing results. I don't think it's unreasonable to ask folks reproducing to use the source version instead of the package.

Am I missing any main cases? Is this how you envision using onir as a package?

If this is the case, I don't think we really need to support the config files in the package environment. (Aside from making sure it only parses command line arguments and such when scripts are run in the development setting.)

**Comment by cash** [comment]
We intend to use the reranking code. Start from query results that may have come from anserini or some other initial retrieval source. Then run the training code to learn a model and create predictions. Then transition back to our own evaluation code (or use yours). As such, I think I need to construct the config to feed into the components. I'm still working through the pipeline code. I imagine we will create our own pipeline class.

I think this PR has the advantage of not changing running from source which is the majority use case.

**Comment by seanmacavaney** [comment]
So each component has default configs in code that give reasonable values. The config files are mostly for setting an array of values for particular experiments, and for practical ease of not needing to have super long command line arguments.

Given this, I think the approach of leaving the config files out of the package seems reasonable to me.

**Comment by cash** [comment]
@seanmacavaney This is ready for review.

The biggest change is moving the jars and scripts into the onir package.

**Review by seanmacavaney — APPROVED** [review]
lgtm
